pkg/parquetquery/predicate_test.go

```diff
@@ -32,41 +33,153 @@ func (p *mockPredicate) KeepValue(parquet.Value) bool             { p.valCalled
 func (p *mockPredicate) KeepPage(parquet.Page) bool               { p.pageCalled = true; return p.ret }
 func (p *mockPredicate) KeepColumnChunk(parquet.ColumnChunk) bool { p.chunkCalled = true; return p.ret }
 
+type predicateTestCase struct {
+	testName   string
+	writeData  func(w *parquet.Writer) //nolint:all
+	keptChunks int
+	keptPages  int
+	keptValues int
+	predicate  Predicate
+}
+
 func TestSubstringPredicate(t *testing.T) {
+	testCases := []predicateTestCase{
+		{
+			testName:   "all chunks/pages/values inspected",
+			predicate:  NewSubstringPredicate("b"),
+			keptChunks: 1,
+			keptPages:  1,
+			keptValues: 2,
+			writeData: func(w *parquet.Writer) { //nolint:all
+				type String struct {
+					S string `parquet:",dict"`
+				}
+				require.NoError(t, w.Write(&String{"abc"})) // kept
+				require.NoError(t, w.Write(&String{"bcd"})) // kept
+				require.NoError(t, w.Write(&String{"cde"})) // skipped
+			},
+		},
+		{
+			testName:   "dictionary in the page header allows for skipping a page",
+			predicate:  NewSubstringPredicate("x"), // Not present in any values
+			keptChunks: 1,
+			keptPages:  0,
+			keptValues: 0,
+			writeData: func(w *parquet.Writer) { //nolint:all
+				type dictString struct {
```

Comment:
Since this struct is now repeated in several cases, let's move it out and maybe rename it `testDictString`.  The key is that it is dictionary encoded which allows the predicate to skip pages.  Not sure what happened because now both `String` and `dictString` are dictionary encoded, but it's ok and let's just use `testDictString` for all 6 cases.